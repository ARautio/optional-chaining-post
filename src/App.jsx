import React from 'react'

import Container from './components/Container'
import Text from './components/Text'

/**
 * Sidebar content list is done by deep object to show how optional chaining works.
 * This certainly isn't the only way to do a sidebar.
 */
const sidebarLinks = {
  FrontPage: {
    Dashboard: ['Most used items', 'Recently added items'],
    'Recent News': ['News from developer', 'News from community']
  },
  Items: {
    'List Items': ['By Date', 'By Location'],
    'Edit Items': []
  },
  Logout: {}
}

const App = () => (
  <Container>
    {Object.keys(sidebarLinks).map(item => (
      <Container key={item}>
        <Text>{item}</Text>
        {/*
         * Optional chaining shortens writing in here since we don't have to check each
         * path level whether it exists or not.
         */}
        {Object.keys(sidebarLinks?.[item] || []).map(subItem => (
          <Container key={subItem}>
            <Text>{subItem}</Text>
            {(sidebarLinks?.[item]?.[subItem] || []).map(subSubItem => (
              <Container key={subSubItem}>
                <Text>{subSubItem}</Text>
              </Container>
            ))}
          </Container>
        ))}
      </Container>
    ))}
  </Container>
)

export default App
