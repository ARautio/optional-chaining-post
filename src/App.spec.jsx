import React from 'react'
import renderer from 'react-test-renderer'

import App from './App'

describe('App', () => {
  it('App creates list of items based on given object', () => {
    expect(renderer.create(<App />).toJSON()).toMatchSnapshot()
  })
})
