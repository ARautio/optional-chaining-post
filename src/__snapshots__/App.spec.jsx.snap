// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App App creates list of items based on given object 1`] = `
<div
  style={
    Object {
      "paddingLeft": "4px",
    }
  }
>
  <div
    style={
      Object {
        "paddingLeft": "4px",
      }
    }
  >
    <p>
      FrontPage
    </p>
    <div
      style={
        Object {
          "paddingLeft": "4px",
        }
      }
    >
      <p>
        Dashboard
      </p>
      <div
        style={
          Object {
            "paddingLeft": "4px",
          }
        }
      >
        <p>
          Most used items
        </p>
      </div>
      <div
        style={
          Object {
            "paddingLeft": "4px",
          }
        }
      >
        <p>
          Recently added items
        </p>
      </div>
    </div>
    <div
      style={
        Object {
          "paddingLeft": "4px",
        }
      }
    >
      <p>
        Recent News
      </p>
      <div
        style={
          Object {
            "paddingLeft": "4px",
          }
        }
      >
        <p>
          News from developer
        </p>
      </div>
      <div
        style={
          Object {
            "paddingLeft": "4px",
          }
        }
      >
        <p>
          News from community
        </p>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "paddingLeft": "4px",
      }
    }
  >
    <p>
      Items
    </p>
    <div
      style={
        Object {
          "paddingLeft": "4px",
        }
      }
    >
      <p>
        List Items
      </p>
      <div
        style={
          Object {
            "paddingLeft": "4px",
          }
        }
      >
        <p>
          By Date
        </p>
      </div>
      <div
        style={
          Object {
            "paddingLeft": "4px",
          }
        }
      >
        <p>
          By Location
        </p>
      </div>
    </div>
    <div
      style={
        Object {
          "paddingLeft": "4px",
        }
      }
    >
      <p>
        Edit Items
      </p>
    </div>
  </div>
  <div
    style={
      Object {
        "paddingLeft": "4px",
      }
    }
  >
    <p>
      Logout
    </p>
  </div>
</div>
`;
